pkg/apis/resource/types.go

```diff
@@ -37,90 +307,335 @@ type ResourceClaim struct {
-	// ResourceClassName references the driver and additional parameters
-	// via the name of a ResourceClass that was created as part of the
-	// driver deployment.
-	ResourceClassName string
+	// Devices defines how to request devices.
+	//
+	// +optional
+	Devices DeviceClaim
 
-	// ParametersRef references a separate object with arbitrary parameters
-	// that will be used by the driver when allocating a resource for the
-	// claim.
+	// Controller is the name of the DRA driver that is meant
+	// to handle allocation of this claim. If empty, allocation is handled
+	// by the scheduler while scheduling a pod.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
 	//
-	// The object must be in the same namespace as the ResourceClaim.
 	// +optional
-	ParametersRef *ResourceClaimParametersReference
+	Controller string
+}
 
-	// Allocation can start immediately or when a Pod wants to use the
-	// resource. "WaitForFirstConsumer" is the default.
+// DeviceClaim defines how to request devices with a ResourceClaim.
+type DeviceClaim struct {
+	// Requests represent individual requests for distinct devices which
+	// must all be satisfied. If empty, nothing needs to be allocated.
+	//
 	// +optional
-	AllocationMode AllocationMode
+	// +listType=atomic
+	Requests []DeviceRequest
+
+	// These constraints must be satisfied by the set of devices that get
+	// allocated for the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Constraints []DeviceConstraint
+
+	// This field holds configuration for multiple potential drivers which
+	// could satisfy requests in this claim. It is ignored while allocating
+	// the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []DeviceClaimConfiguration
 }
 
-// AllocationMode describes whether a ResourceClaim gets allocated immediately
-// when it gets created (AllocationModeImmediate) or whether allocation is
-// delayed until it is needed for a Pod
-// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
-// future.
-type AllocationMode string
+// DeviceRequest is a request for devices required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices.
+type DeviceRequest struct {
+	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
+	// entry and in a constraint of the claim.
+	//
+	// Must be a DNS label.
+	//
+	// +required
+	Name string
 
+	*DeviceRequestDetails // inline
+}
+
+// DeviceRequestDetails is embedded inside DeviceRequest. Exactly one field must be set.
+type DeviceRequestDetails struct {
+	// Device requests one or more devices.
+	//
+	// +optional
+	Device *DeviceRequestDetail
+}
+
+// DeviceRequestDetail is currently the only permitted alternative in DeviceRequestDetails.
+type DeviceRequestDetail struct {
```

Comment:
Should this (or it's wrapping struct) have a different name? It feels weird to me to have a "pluralized" struct with a single field that is then the "non-pluralized" version of it. Especially if that single field is actually the first in a list of future "one-ofs".

This would go away if we didn't do the inlining, but in lieu of that, maybe it would make it easier to come up with a different name if I understood what alternatives for `DeviceRequestDetail` you envision in the future.

Reply:
Let's discuss in https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1664073074.